### PR TITLE
Server-Timing HTTP header field

### PR DESCRIPTION
**Pull request by alex-schneider**
Describe your changes!

---
<details>
    <summary>Reviewer checklist</summary>
    <ul>
        <li>Read PR description: a summary about the changes is required</li>
        <li>Changelog updated</li>
        <li>Documentation: docs/{Reference, Cli, ...}, Docker and cli help/usage</li>
        <li>Pulled branch, manually tested</li>
        <li>Verified requirements are met</li>
        <li>Reviewed the code</li>
        <li>Reviewed the related tests</li>
    </ul>
</details>

### CODE REVIEW

**Comment by johakoch** [comment]
`make generate-docs`

File: logging/upstream_log.go [hunk 1]
```diff
@@ -149,17 +152,37 @@ func (u *UpstreamLog) RoundTrip(req *http.Request) (*http.Response, error) {
...
+			serverTimingsKey = key.(string)
+		}
+	}
+
+	if name, ok := requestFields["name"].(string); ok && name != "default" {
+		serverTimingsKey += "_" + name
+	}
+
 	timingResults := Fields{
 		"total": RoundMS(rtDone.Sub(rtStart)),
```
Comment (johakoch) [inline logging/upstream_log.go:170]: Also add total to serverTimingsVal.

File: utils/server_timing.go [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+
+		if _, exists := timings[newKey]; !exists {
+			return newKey
+		}
+
+		i++
+	}
+}
+
+// RFC 2616
```
Comment (johakoch) [inline utils/server_timing.go:117]: https://httpwg.org/specs/rfc7230.html#rule.token.separators
The resulting valid chars for production `token` are the same.

File: utils/server_timing.go [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+func isTokenChar(ch byte) bool {
+	const separators = `()<>@,;:\"/[]?={}` + " " + "\t"
```
Comment (johakoch) [inline utils/server_timing.go:119]: You can drop `\t`, as it is within `ch <= 31`.

File: utils/server_timing.go [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+
+	if ch <= 31 || ch == 127 {
```
Comment (johakoch) [inline utils/server_timing.go:121]: If you write `ch <= 32`, you can also drop `" "` (x20 = 32) from `separators`.

File: logging/upstream_log.go [hunk 5]
```diff
@@ -149,17 +152,37 @@ func (u *UpstreamLog) RoundTrip(req *http.Request) (*http.Response, error) {
...
+	serverTimingsKey := ""
+
+	// For test cases: "backend" is sometimes not set
+	if u.log != nil {
+		if key, exists := u.log.Data["backend"]; exists {
+			serverTimingsKey = key.(string)
+		}
+	}
+
+	if name, ok := requestFields["name"].(string); ok && name != "default" {
```
Comment (johakoch) [inline logging/upstream_log.go:165]: `configload.defaultNameLabel`, just in case, the default name is changed :-)

File: logging/upstream_log.go [hunk 6]
```diff
@@ -149,17 +152,37 @@ func (u *UpstreamLog) RoundTrip(req *http.Request) (*http.Response, error) {
...
+		serverTimingsKey += "_" + name
```
Comment (johakoch) [inline logging/upstream_log.go:166]: Should we use different separators for adding
* request name
* timing name
* metric distinguisher (1, 2, ...)

? Other valid separators could be `.` or `-`.

File: utils/server_timing.go (hunk 2 above)
Comment (alex-schneider) [reply utils/server_timing.go:117]: Updated...

File: logging/upstream_log.go (hunk 5 above)
Comment (alex-schneider) [reply logging/upstream_log.go:165]: Done...

File: utils/server_timing.go (hunk 3 above)
Comment (alex-schneider) [reply utils/server_timing.go:119]: I would like to leave the list of separators completely, like defined in RFC :)

File: utils/server_timing.go (hunk 4 above)
Comment (alex-schneider) [reply utils/server_timing.go:121]: I would like to leave the list of separators completely, like defined in RFC :)

File: logging/upstream_log.go (hunk 1 above)
Comment (alex-schneider) [reply logging/upstream_log.go:170]: Done...

File: logging/upstream_log.go (hunk 6 above)
Comment (alex-schneider) [reply logging/upstream_log.go:167]: @malud , @filex 

What do you mean?

Comment (filex) [reply logging/upstream_log.go:167]: The header is informational/ for debugging. I'd keep it simple and stick to _. Users will still be able to identify which metric came from which request.

(Another aspect would be accessing the metrics from JS via the ServerTiming API. Here, I like underscores, too. But I don't think that anyone will actually go that way.)

File: utils/server_timing.go (hunk 3 above)
Comment (johakoch) [reply utils/server_timing.go:119]: As you now reference the current RFC (BTW, I would rather add the URL that I provided (https://httpwg.org/specs/rfc7230.html#rule.token.separators) or at least the section number), which defines token in a _positive_ way (unlike RFC 2616), this argument does not count anymore.

File: config/settings.go [hunk 7]
```diff
@@ -50,6 +51,7 @@ type Settings struct {
 	RequestIDClientHeader     string `hcl:"request_id_client_header,optional" docs:"HTTP header field which Couper uses to transport the {request.id} to the client" default:"Couper-Request-ID"`
 	RequestIDFormat           string `hcl:"request_id_format,optional" docs:"{\"common\"} or {\"uuid4\"}. If set to {\"uuid4\"} an RFC 4122 UUID is used for {request.id} and related log fields. " default:"common"`
 	SecureCookies             string `hcl:"secure_cookies,optional" docs:"{\"\"} or {\"strip\"}. If set to {\"strip\"}, the {Secure} flag is removed from all {Set-Cookie} HTTP header fields." default:"\u200C"`
+	SendServerTimings         bool   `hcl:"send_server_timing_headers,optional" docs:"If enabled, Couper includes an additional [Server-Timing](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Server-Timing) HTTP header field to the client detailing connection and transport relevant metrics for each backend request."`
```
Comment (johakoch) [inline config/settings.go:54]:
```suggestion
	SendServerTimings         bool   `hcl:"send_server_timing_headers,optional" docs:"If enabled, Couper includes an additional [Server-Timing](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Server-Timing) HTTP response header field detailing connection and transport relevant metrics for each backend request."`
```

Comment (johakoch) [inline config/settings.go:54]: Also add docu to https://docs.couper.io/configuration/command-line

File: utils/server_timing.go (hunk 3 above)
Comment (alex-schneider) [reply utils/server_timing.go:119]: Done...

File: config/settings.go (hunk 7 above)
Comment (alex-schneider) [reply config/settings.go]: Done...

**Comment by johakoch** [comment]
Currently, the metrics may have a suffix (`_[0-9a-f]{6}`), which is unique, ok, but it's not possible to identify the position in a sequence of couper layers. Better use numbers, again.

Another idea: Instead of suffixes, use prefixes, as Chrome seems to sort Server Timings in the Timing tab.

**Review by johakoch — APPROVED** [review]

**Pull request merged**